src/module/s3.js

```diff
@@ -1,28 +1,63 @@
+const util = require('util');
 const zlib = require('zlib');
+const get = require('lodash.get');
 const Joi = require('joi-strict');
 
-module.exports.S3 = ({ call }) => {
-  const putGzipObject = ({ bucket, key, data }) => call('s3:putObject', {
+const sleep = util.promisify(setTimeout);
+
+module.exports.S3 = ({
+  call,
+  logger,
+  backoffFunction = (count) => 30 * (count ** 2),
+  maxRetries = 10
+}) => {
+  const exec = async (
+    action,
+    opts,
+    { expectedErrorCodes = [] } = {}
+  ) => {
+    let lastError;
+    for (let count = 0; count < maxRetries; count += 1) {
+      // eslint-disable-next-line no-await-in-loop
+      await sleep(backoffFunction(count));
+      try {
+        // eslint-disable-next-line no-await-in-loop
+        return await call(action, opts, { expectedErrorCodes });
+      } catch (e) {
+        lastError = e;
+        if (get(e, 'message') === 'SlowDown: Please reduce your request rate.') {
+          if (logger !== null) {
+            logger.warn(`Failed to submit (some) ${action}(s)\nRetrying: [${JSON.stringify({ action, opts })}]`);
+          }
+        } else {
```

Comment:
I'm still not logging out the retries because there are only two ways here, to get a different error after some retries or reach the max allowed value. So we need to define how that could be useful.